### PR TITLE
[SYCL][libdevice] Tests for half value type cast utils.

### PR DESCRIPTION
**Pull request by jinge90**
Signed-off-by: jinge90 <ge.jin@intel.com>

### CODE REVIEW

**Comment by jinge90** [comment]
This PR is test for https://github.com/intel/llvm/pull/6930

File: SYCL/DeviceLib/imf_double2half.cpp [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+            << device_queue.get_device().get_info<sycl::info::device::name>()
+            << "\n";
+
+  if (!device_queue.get_device().has(sycl::aspect::fp64)) {
+    std::cout << "Test skipped on platform without fp64 support." << std::endl;
+    return 0;
+  }
+
+  {
+    std::initializer_list<uint64_t> input_vals = {
```
Comment (akolesov-nv) [inline SYCL/DeviceLib/imf_double2half.cpp:33]: Is this just a sanity check or detailed test?
The test is good as a sanity one.
For detailed test I would suggest to extend it by:
* all DP corner points: +/-zero, +/-max&min subnormals, +/-max&min normals, +/-infs, nans
* all HP expected output corner points: +/-zero, +/-max&min subnormals, +/-max&min normals
(some of them are already covered)
* + all possible exponent fields (with random mantissa)

File: SYCL/DeviceLib/imf_half_type_cast.cpp [hunk 2]
```diff
@@ -0,0 +1,589 @@
...
+    test_host(input_vals, ref_vals_ru, F_Half1(__imf_half2ull_ru));
+    test(device_queue, input_vals, ref_vals_ru, F_Half1(__imf_half2ull_ru));
+
+    test_host(input_vals, ref_vals_rz, F_Half1(__imf_half2ull_rz));
+    test(device_queue, input_vals, ref_vals_rz, F_Half1(__imf_half2ull_rz));
+  }
+
+  {
+    std::initializer_list<short> input_vals = {
+        0, 7, 49, -323, 2999, -32768, 32767, 32111, -11111, 16383};
```
Comment (akolesov-nv) [inline SYCL/DeviceLib/imf_half_type_cast.cpp:465]: General question: do you check integer indefinite (-1.#IND) inputs somewhere?

Comment (jinge90) [reply SYCL/DeviceLib/imf_half_type_cast.cpp:465]: Hi, @akolesov-intel 
What do you mean by "integer indefinite"?

Thanks very much.

Comment (akolesov-nv) [reply SYCL/DeviceLib/imf_half_type_cast.cpp:465]: Sorry, it is not related to conversions actually, I was wrong. I was thinking about exceptions handling

**Review by akolesov-nv — APPROVED** [review]

File: SYCL/DeviceLib/imf_double2half.cpp (hunk 1 above)
Comment (jinge90) [reply SYCL/DeviceLib/imf_double2half.cpp:33]: Hi, @akolesov-intel 
This test is a lit test which is similar to sanity test. And I also added +0/-0, +/-max & min subnormals, +/- max & min normals, +/-infs
Thanks very much.

**Comment by jinge90** [comment]
Hi, @steffenlarsen and @intel/llvm-reviewers-runtime 
There is one failure in pre-ci which is xpti/content.cpp, this failure is not related to this PR. Could you help review?

Thanks very much.

**Review by akolesov-nv — APPROVED** [review]

**Comment by jinge90** [comment]
Hi, @steffenlarsen and @intel/llvm-reviewers-runtime 
Kind ping~

File: SYCL/DeviceLib/imf_double2half.cpp [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+// RUN: %CPU_RUN_PLACEHOLDER %t.out
+// RUN: %ACC_RUN_PLACEHOLDER %t.out
+// RUN: %GPU_RUN_PLACEHOLDER %t.out
+//
+// UNSUPPORTED: cuda || hip
+
+#include "imf_utils.hpp"
+
+extern "C" {
+_iml_half_internal __imf_double2half(double);
```
Comment (steffenlarsen) [inline SYCL/DeviceLib/imf_double2half.cpp:12]: Is `_iml_half_internal` always supported or should this test also require that the corresponding device has `sycl::aspect::fp16`? Same goes for the other tests.

Comment (jinge90) [reply SYCL/DeviceLib/imf_double2half.cpp:12]: Hi, @steffenlarsen 
Thanks for catching this, already updated the patch.

Thanks very much.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Comment by steffenlarsen** [comment]
Failures are unrelated.

Linux Level-Zero failure:
SYCL :: Plugin/level_zero_interop_memcpy.cpp

Windows OpenCL failure:
SYCL :: KernelAndProgram/kernel-bundle-merge-options-env.cpp

**Pull request merged**